### ISSUE TITLE
ruby cucumber-expressions yields arity exception with optional ParameterType args

### ISSUE THREAD

**Issue by mwhipple** [cucumber/cucumber#761]
## Summary

`ArgumentError`s are thrown when trying to invoke a Transformer function for a ParameterType when some of the arguments are missing.

## Expected Behavior

The transformer function should be invoked with the proper number of parameters where any omitted in the capture are passed as `nil`. This would be highly preferable to shuffling parameters based on presence and appears to be the expected behavior based on the _Nested capture groups_ example at: https://cucumber.io/blog/upgrading-to-cucumber-3/

## Current Behavior

In cases where a ParameterType includes a regex with an optional capture and no value is provided for that capture then an error such as:
`wrong number of arguments (given 0, expected 2) (ArgumentError)` is thrown.

## Possible Solution

Some initial digging indicates that this may have been introduced by the addition of `compact` on this line:
https://github.com/cucumber/cucumber/blame/master/cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/group.rb#L14

in this commit:
https://github.com/cucumber/cucumber/commit/c12b9c6e5a8d22d589c455f3aa819fa795ffcc22

## Steps to Reproduce (for bugs)

1. Define a ParameterType with a regexp such as ```/(?: child(ren)? `([^`]*)`)?/```
2. Add a `transformer` to the ParamterType without specifying optional parameters.
2. Write a spec which uses the ParameterType and omits optional text.

## Context & Motivation

This works with optional parameters, however the inability to rely on expected positional parameters results in a fair amount of noise in the logic and a degraded experience from Cucumber 2.

## Your Environment

* Version used: cucumber 3.1.2; ruby 2.5.7
* Operating System and version: Linux 5.1.8 (Gentoo)
* Link to your project: https://github.com/brightcove/brine/

**Comment by luke-hill** [comment]
Just to clarify, as your initial triage is a bit wordy. Is the issue the following.

`/(?: child(ren)? `([^`]*)`)?/` -> Has captures of
1 => `"ren"` or `nil`
2 => a string of indeterminate length.

And This entire thing is wrapped in a non-capturing capture which might be optional. So both of the above could be optional.

Whereas you want the number of captures to be 0, when the `?` optional is zero. And as such yield 0 captures.

Or have I misunderstood.

Could you write a small feature test to pull from a git repo?

**Comment by aslakhellesoy** [comment]
> Could you write a small feature test to pull from a git repo?

Or even better (and simpler!): A pull request with a failing unit test.

**Comment by mwhipple** [comment]
I meant to include a concrete example also, sorry. The example parameter bindings you've provided look right with the additional mentioned case where the entire clause is absent and both parameters are passed through as `nil`. The specific annoyance is only around where any omitted arguments precede those that are provided: for example the 
`Given child bobby, sue => 1 => nil, 2=> bobby, sue` case in the table above above. For the others using `->(plurality=nil, child_list=nil)` doesn't require any additional parameter sniffing in the body since the provided sparse arguments would fill in the right parameters in order. I'm running in to the behavior consistently (for much simpler regexes) but the others are easily worked around with default param values.

I'll work on getting a test together this week (or smoking out any issues on my end that may be causing this).

**Comment by aslakhellesoy** [comment]
Let’s leave this open a little longer

**Comment by mwhipple** [comment]
I'll pick this back up this week; I got distracted with a job change.

**Comment by mwhipple** [comment]
A fairly minimal reproduction was opened as a failing RSpec test; it if looks like a legitimate issue I'll work to get the tests fixed up/approved. If the **Possible Solution** above sounds feasible or there's any suggested approaches I can probably look at resolving it but I'll be unlikely to have near term time to do deeper digging.

**Comment by luke-hill** [comment]
We can probably close this @mpkorstanje over the PR in the cucumber-messages repo.

On a side-note @mwhipple - As you submitted this around xmas I lost it. I'll get onto it next week! Very easy to reproduce.

**Comment by mwhipple** [comment]
Is #771 the work alluded to that could likely incidentally resolve this?

**Comment by mpkorstanje** [comment]
@mwhipple I don't think so. While #771 touches on cucmber expression parsing, it doesn't touch parsing the regular expressions in parameter types.

@luke-hill could you clarify?

**Comment by mwhipple** [comment]
I likely just connected dots that weren't there so I should step back. Is this issue still a good place to track this or does it overlap with something else? Is #838 worth keeping/cleaning up or should it be closed?

**Comment by luke-hill** [comment]
So the reproducible test gives me enough info to triage this. It's quite simple what's going wrong. Obviously I'm not saying the solution will be easy.

As/when I have time I can triage. The PR he created to showcase the issue gives more info than this, so lets close this over the PR.

**Comment by mpkorstanje** [comment]
Superceded by #838
